extra/mariadb_upgrade_config_file.c

```diff
@@ -0,0 +1,1178 @@
+                orig, name, strerror(errno));
+        my_free(orig);
+        return 1;
+      }
+    }
+    my_free(orig);
+    my_free(name);
+  }
+  delete_dynamic(&ctx->updated_files);
+  return 0;
+}
+
+
+/**
+  Run a command using the shell, storing its output in the supplied dynamic
+  string.
+*/
+static int run_command(char* cmd,
+                       DYNAMIC_STRING *ds_res)
+{
+  char buf[512]= {0};
+  FILE *res_file;
+  int error;
+
+  if (!(res_file= my_popen(cmd, "r")))
+  {
+    fprintf(stderr, "popen(\"%s\", \"r\") failed\n", cmd);
+    return -1;
+  }
+
+  while (fgets(buf, sizeof(buf), res_file))
+  {
+#ifdef _WIN32
+    /* Strip '\r' off newlines. */
+    size_t len = strlen(buf);
+    if (len > 1 && buf[len - 2] == '\r' && buf[len - 1] == '\n')
+    {
+      buf[len - 2] = '\n';
+      buf[len - 1] = 0;
+    }
+#endif
+    dynstr_append(ds_res, buf);
+  }
+
+  error= my_pclose(res_file);
+  return WEXITSTATUS(error);
+}
+
+
+/**
+  Run `mariadbd --help --verbose` with the supplied arguments and write its
+  stderr output to ds_res.
+*/
+static int run_mariadbd(const char *mariadbd_path,
+                        DYNAMIC_STRING *ds_res,
+                        const char **defaults_args,
+                        int defaults_args_count)
+{
+  int ret;
+  int i;
+  DYNAMIC_STRING ds_cmdline;
+
+  if (init_dynamic_string(&ds_cmdline, IF_WIN("\"", ""), FN_REFLEN, FN_REFLEN))
+  {
+    fputs("Out of memory\n", stderr);
+    return -1;
+  }
+
+  dynstr_append_os_quoted(&ds_cmdline, mariadbd_path, NullS);
+  dynstr_append(&ds_cmdline, " ");
+
+  for (i= 0; i < defaults_args_count; i++)
+  {
+    dynstr_append_os_quoted(&ds_cmdline, defaults_args[i], NullS);
+    dynstr_append(&ds_cmdline, " ");
+  }
+
+  dynstr_append(&ds_cmdline, "--help ");
+  dynstr_append(&ds_cmdline, "--verbose ");
+  dynstr_append(&ds_cmdline, "2>&1 ");
+  dynstr_append(&ds_cmdline, IF_WIN("1>NUL", "1>/dev/null"));
+
+#ifdef _WIN32
+  dynstr_append(&ds_cmdline, "\"");
+#endif
+
+  ret= run_command(ds_cmdline.str, ds_res);
+  dynstr_free(&ds_cmdline);
+  return ret;
+}
+
+
+/**
+  Test whether mariadbd can be launched with --no-defaults.
+*/
+static int test_mariadbd(const char *mariadbd_name)
+{
+  DYNAMIC_STRING ds_tmp;
+  const char *defaults_argument = "--no-defaults";
+  int err= 0;
+
+  if (init_dynamic_string(&ds_tmp, "", 32, 32))
+  {
+    fputs("Out of memory\n", stderr);
+    return -1;
+  }
+
+  if (run_mariadbd(mariadbd_name,
+               &ds_tmp,
+               &defaults_argument,
+               1))
+  {
+    fprintf(stderr, "Can't execute %s\n", mariadbd_name);
+    err= 1;
+  }
+
+  dynstr_free(&ds_tmp);
+  return err;
+}
+
+
+static int compare_options(const void *a, const void *b)
+{
+  const char *first= *(const char **)a;
+  const char *second= *(const char **)b;
+  return strcmp(first, second);
+}
+
+
+static my_bool mariadbd_option_exists(const char *option)
+{
+   return bsearch(&option, mariadbd_valid_options,
```

Comment:
Can you sort these options before you apply bsearch, in the generator program?

Apparently, on Windows it produces something that is not sorted, neither are the options unique.
This is what I got (after removing --plugin-load-add, for mariadbd --help)
```
static const char *mariadbd_valid_options[]= {
"install",
"install_manual",
"install",
"install_manual",
"remove",
"remove",
"enable_named_pipe",
"standalone",
"allow_suspicious_udfs",
"alter_algorithm",
"analyze_sample_percentage",
"aria_block_size",
"aria_checkpoint_interval",
```
After `allow_suspicious_udf` , it seems sorted,  but before it there is some Windows specific stuff,  random order, with repetitions.

Reply:
Apparently we may [produce an empty array](https://buildbot.mariadb.org/#/builders/234/builds/26156):
```
         C:\Buildbot\amd64-windows\build\extra\mariadb_upgrade_config_file.c(242,12): warning C4034: sizeof returns 0 [C:\Buildbot\amd64-windows\build\extra\mariadb-upgrade-config-file.vcxproj]
       C:\Buildbot\amd64-windows\build\extra\mariadb_upgrade_config_file.c(242,12): warning C4034:            array_elements(mariadbd_valid_options), [C:\Buildbot\amd64-windows\build\extra\mariadb-upgrade-config-file.vcxproj]
       C:\Buildbot\amd64-windows\build\extra\mariadb_upgrade_config_file.c(242,12): warning C4034:            ^ [C:\Buildbot\amd64-windows\build\extra\mariadb-upgrade-config-file.vcxproj]
```